### ISSUE TITLE
Invalid initializer shader compile error 

### ISSUE THREAD

**Issue by ratsdiov** [godotengine/godot#2541]
I've been converting several GLSL shaders to Godot and hit a few strange errors that change if I move code around.  Here is one I was able to replicate in a concise example.  If the foobar function simply exists in the file it causes the unrelated test function to cause a compile error.  Also just swapping the order of foobar and test makes it work ok.   Very difficult to debug and trace these errors down but I'll try to submit other examples as I can.  I can send a zip of this test case as a project by email if desired.

I'm using 1.1 Stable under windows 7 64 bit.

```
// Uncomment the following and the shader fails to compile with error:
// "error(1,0): Invalid initializer for variable: testR"
// Also if test is defined before foobar it *will* work ok
/*
float foobar(vec2 co){
    return 1.0;
}
*/

vec2 test(vec2 p)
{
	return p;
}

vec2 testR = test(UV);
COLOR = color(testR, 0.0, 1.0);
```

**Comment by ratsdiov** [comment]
Here is another... these are not fun to isolate.  Hopefully these two examples will help pinpoint the problem.

```

vec3 test_3(vec3 x) {
  return vec3(0.0);
}
// Commenting out test_2 makes it work, with test_2 present it fails with:
// "error(1,0): Invalid initializer for variable: n"
vec2 test_2(vec2 x) {
  return vec2(0.0);
}

float test(vec2 v)
{
	return .5;
}

float n = test(UV);
COLOR = color(n);
COLOR.a = 1.0;

```

**Comment by Magurosenbei** [comment]
This is a bug due to broken logic within validate_function_call(....) in 
shader_language.cpp: Line 1373
As the parser identifies a function only by its arguments types but not its name.
So if you have a function that is
type1 func1 (arg1, arg2 ... argN)
type2 func2 (arg1, arg2 ... argN)

a statement such as
type2 test = func2(arg1...argN)
will be validated using func1, resulting in its return cache being of type1 instead of type2.

Fix this in shader_language.cpp Line 1373 by changing from  
`if (!fail)`  to `if (!fail && name == program->functions[i].name)`

**Comment by akien-mga** [comment]
Thanks for the debugging! Could you make a pull request with your proposed fix?

**Comment by Magurosenbei** [comment]
Done.

**Comment by akien-mga** [comment]
Fixed by #2926.
